netbox/dcim/signals.py

```diff
@@ -123,3 +125,15 @@ def nullify_connected_endpoints(instance, **kwargs):
 
     for cablepath in CablePath.objects.filter(_nodes__contains=instance.cable):
         cablepath.retrace()
+
+
+@receiver(m2m_changed, sender=Interface.vdcs.through)
+def enforce_vdc_type_restrictions(instance, **kwargs):
+    if 'action' == 'post_add':
+        device = instance.device
+        if device.device_type.vdc_type not in [VirtualDeviceContextTypeChoices.CISCO_ASA_CONTEXT, VirtualDeviceContextTypeChoices.CISCO_FTD_INSTANCE] \
+                and len(instance.vdcs) > 1:
```

Comment:
This sort of validation should take place in the model's `clean()` method.

Reply:
M2M set operations are performed after the model's save(), so it wouldn't really hit the mdoels clean() method either.  You would see old data, but if you are adding a interface it would bypass the check.

We could remove this, as the form clean should catch it, we would have to do something different for the API however.